### ISSUE TITLE
get eax value in hook cpuid

### ISSUE THREAD

**Issue by PhilippTakacs** [unicorn-engine/unicorn#1516]
As a workaround for #1515 I tried to use the hook cpuid. The problem is that the hook is called after cpu_x86_cpuid() is called and the registers are updated. So you can't update the registers depending on eax.

I see two ways to fix this:

* call the hook before cpu_x86_cpuid() and get a boolean if cpu_x86_cpuid() should be called.
* add an eax parameter to the hook

**Comment by wtdcode** [comment]
Hello, I don't understand why you can't update the registers in the hook?

**Comment by PhilippTakacs** [comment]
I have some sample code:

```
#define OK(x) {uc_err __err; if ((__err = x)) { fprintf(stderr, "%s", uc_strerror(__err)); assert(false); } }
void cpuid_callback(uc_engine *uc, void *data)
{
    int reg = 7;

    OK(uc_reg_read(uc, UC_X86_REG_EAX, &reg));
    if (reg == 1) {
        reg = 7;
        OK(uc_reg_write(uc, UC_X86_REG_EAX, &reg));
    } else {
        reg = 8;
        OK(uc_reg_write(uc, UC_X86_REG_EAX, &reg));
    }
}


static void test_x86_hook_cpuid()
{
    uc_engine *uc;
    char code[] = "\xB8\x01\x00\x00\x00\x0F\xA2"; // MOV 0x01 EAX; CPUID
    uc_hook h;
    int reg;

    OK(uc_open(UC_ARCH_X86, UC_MODE_32, &uc));
    OK(uc_mem_map(uc, 0, 4 * 1024, UC_PROT_ALL));
    OK(uc_mem_write(uc, 0, code, sizeof(code)-1));


    OK(uc_hook_add(uc, &h, UC_HOOK_INSN, cpuid_callback, NULL, 1,
                   0, UC_X86_INS_CPUID));

    OK(uc_emu_start(uc, 0, sizeof(code) - 1, 0, 0));

    OK(uc_reg_read(uc, UC_X86_REG_EAX, &reg));

    assert(reg == 7);

    OK(uc_close(uc));
}
```

The problem is cpuid_callback() is called after eax is updated https://github.com/unicorn-engine/unicorn/blob/63a445cbba18bf1313ac3699b5d25462b5d529f4/qemu/target-i386/misc_helper.c#L84  So I can't change the value depending on what the code has eax set to.

**Comment by wtdcode** [comment]
Why do you depend on EAX to update CPUID?

**Comment by PhilippTakacs** [comment]
Because CPUID depends on EAX. Without knowing what EAX was before the call a hook can't determine what to change. I have digged a bit in the code and found CPUID depends also on ECX. So having the old ECX availible would also be usefull.

**Comment by wtdcode** [comment]
Fixed in 7bb756249a54a2163d5cb65f35b64912451f0752
